### PR TITLE
Titan: Show post-checkout upsell for eligible domains

### PR DESCRIPTION
**Pull request by stephanethomas**
This pull request is a follow-up of https://github.com/Automattic/wp-calypso/pull/56671 and https://github.com/Automattic/wp-calypso/pull/56690. It updates the purchase flow to show an upsell for Professional Email after the checkout under certain conditions:

screenshot

This pull request:

* Adds an `upsell/professional-email` feature gate
* Adds a function that retrieves the first domain eligible to Professional Email from a list of domains
* Adds a function that retrieves the url of the Professional Email upsell page
* Adds a hook that can be used to retrieve a list of domains for a site
* Renames several methods dealing with upsells to follow the same naming convention
* Updates the logic that determines which upsell should be shown to a user
* Adds unit tests for those changes

Note the logic for showing and handling an upsell page follows these high-level steps:

1. Determines what page should be displayed when purchase is successful
2. Redirects the user to that page
3. Shows the upsell
4. Determines what page should be displayed when an upsell is declined

That means we need to retrieve a domain eligible to Professional Email in step #1 but also in step #3 (we don't pass the domain name in the url of the Professional Email upsell page). A domain is eligible if it:

* Has not expired
* Is not a staging domain (Atomic)
* Has neither Titan nor Google Workspace already
* Can be added/managed by the user

In addition to that, the Professional Email upsell page should only be displayed when purchasing the Blogger, Personal, or Business plans. Right now this page is also managed by the `upsell/professional-email` feature gate. We will move to an A/B test in another pull request.

#### Testing instructions

You can run unit tests with:

```javascript
npm run test-client client/lib/titan
npm run test-client client/my-sites/checkout/composite-checkout
```

Otherwise you can:

1. Run `git checkout update/professional-email-upsell` and start your server, or access a [live branch](https://github.com/Automattic/wp-calypso/pull/56794#issuecomment-934563437)
2. Paste the following in your browser's console to debug the hook:
```javascript
localStorage.setItem( 'debug', 'calypso:composite-checkout:use-site-domains' )
```
3. Log into a WordPress.com account that has a site with an eligible domain
4. Open the [`Plans` page](http://calypso.localhost:3000/plans)
5. Purchase any plan except Premium
6. Assert that you are presented with the Professional Email upsell
7. Assert that that page still works if you refresh it
8. Upgrade your plan to Premium
9. Assert that you are still presented with the Business Plan upsell
10. Upgrade your plan to Business
11. Assert that you are presented with the Professional Email upsell
12. Navigate to the [`Emails` page](http://calypso.localhost:3000/email/stephanethomas494.wordpress.com)
13. Click the `Add Email` button located in front of your eligible domain
14. Add either Titan or Google Workspace account to your shopping cart (but don't purchase it)
15. Head back to the `Plans` page
16. Click the `Upgrade` button below the eCommerce plan
17. Assert that the `Checkout` page shows both that plan and the email solution you chose before
18. Pay, and assert that you are not presented with any upsell (you are taken to the site setup flow)

You may want to follow these [instructions](https://github.com/Automattic/wp-calypso/pull/43199) to cover existing use cases.

### CODE REVIEW

File: client/lib/titan/is-domain-eligible-for-titan-free-trial.js [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/**
+ * Determines if the specified domain is eligible to the Titan 3-month free trial.
+ *
+ * @param {object} domain - domain object
+ * @returns {?boolean} whether the domain is eligible or not, undefined otherwise
+ */
+export function isDomainEligibleForTitanFreeTrial( domain ) {
```
Comment (daledupreez) [inline client/lib/titan/is-domain-eligible-for-titan-free-trial.js]: Thought: it may be worth pushing this refactor into a separate PR to make it easier to test everything -- the test for that PR should be very simple and low-risk.

File: client/lib/titan/get-eligible-titan-domain.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	if ( ! domains ) {
+		return null;
+	}
+
+	const eligibleDomains = domains.filter( ( domain ) => {
+		if ( domain.expired || domain.isWPCOMDomain ) {
+			return false;
+		}
+
+		if ( hasTitanMailWithUs( domain ) ) {
```
Comment (daledupreez) [inline client/lib/titan/get-eligible-titan-domain.js]: Should we also check for `hasGSuiteWithUs( domain )` as well? Maybe that should be an option to `getEligibleTitanDomain()`?

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts [hunk 3]
```diff
@@ -11,6 +11,7 @@ import {
 	isPlan,
 	isWpComPremiumPlan,
 } from '@automattic/calypso-products';
+import { Domain } from '@automattic/data-stores';
```
Comment (daledupreez) [inline client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts]: Should this not be `import type`?

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts [hunk 4]
```diff
@@ -408,25 +421,37 @@ function maybeShowPlanBumpOffer( {
...
+		pendingOrReceiptId,
+		cart,
+		orderId,
+		siteSlug,
+		domains,
+	} );
+
+	if ( professionalEmailUpsellUrl ) {
+		return professionalEmailUpsellUrl;
+	}
```
Comment (daledupreez) [inline client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:454]: I think we need to inject the relevant experiment here, likely via the `useExperiment` hook (which is probably not directly available from this context). I think we do want to confirm that the user is eligible for the A/B test before we use the experiment assignment to return either the Professional Email or Quick Start upsell URL. (More details are available in this README: https://github.com/Automattic/wp-calypso/blob/trunk/packages/explat-client-react-helpers/README.md.)

From a logical perspective, I think we want to do something like the following:
```
const isUserEligibleForProfessionalEmailUpsell = ! hasTitanMail( cart ) && ! hasPremiumPlan( cart )
  && ( hasBloggerPlan( cart ) /* we don't sell the blogger plan any more */ || hasPersonalPlan( cart ) || hasBusinessPlan( cart ) /* || hasECommercePlan( cart ) - not sure if this exists */ )
  && getEligibleTitanDomain( siteSlug, domains ); /* TODO: ensure the domain is eligible for the free trial */
// Maybe log a tracks event so we know what proportion of users are eligible for this offer, and possibly why not
if ( isUserEligibleForProfessionalEmailUpsell ) {
  const [ isLoadingExperimentAssignment, experimentAssignment ] = useExperiment( 'promote_professional_email_post_checkout_2021_10' );
  if ( isLoadingExperimentAssignment ) {
    // Not sure how this would work...
  } else if ( experimentAssignment?.variationName === 'treatment' ) {
    // return Professional Email upsell URL
  } else {
    // return Quick Start upsell URL, or possibly no upsell URL
  }
}
```

Comment (daledupreez) [reply client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:454]: A notable side-effect of the change above is that we don't need the feature flag any more -- we should be able to rely on the experiment assignment.

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts [hunk 5]
```diff
@@ -475,9 +502,57 @@ function getQuickstartUrl( {
...
 	}
+
+	// This is used when we need to quickly dial back the volume of concierge sessions
+	// being offered and sold, to be inline with HE availability.
+	// Change this percentage to change the amount of offers given out:
+	const percentConciergeOffers = 75;
+
+	if ( ! badNaiveClientSideRollout( 'conciergeUpsellDial', percentConciergeOffers ) ) {
+		return;
+	}
```
Comment (daledupreez) [inline client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:514]: Hmmm... I was thinking we'd simply be in a position to switch over to the A/B test, but I now realise that we need some interim handling for the time before we launch the experiment, otherwise we're going to be showing the Quick Start upsell to all users who are eligible for the Professional Email upsell prior to launching the experiment.

I am wondering if removing this upsell may be the right starting point.

File: client/lib/titan/get-eligible-titan-domain.js (hunk 2 above)
Comment (AllTerrainDeveloper) [reply client/lib/titan/get-eligible-titan-domain.js]: Good point, makes sense to me. A domain should not be elegible to sell an email provider if already has a Google Workspace on it.

File: client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js [hunk 6]
```diff
@@ -6,9 +6,12 @@
 
 import { isEnabled } from '@automattic/calypso-config';
 import {
+	PLAN_BLOGGER,
 	PLAN_ECOMMERCE,
 	JETPACK_REDIRECT_URL,
+	PLAN_PREMIUM,
 	redirectCheckoutToWpAdmin,
+	TITAN_MAIL_MONTHLY_SLUG,
 	WPCOM_DIFM_LITE,
```
Comment (AllTerrainDeveloper) [inline client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js]: Import order? :) Anyway I guess that's ok. And by the way, nice tests!!

**Comment by stephanethomas** [comment]
A heads-up that I've just rebased that branch against `trunk`.

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts (hunk 3 above)
Comment (stephanethomas) [reply client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts]: Good catch!

File: client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js (hunk 6 above)
Comment (stephanethomas) [reply client/my-sites/checkout/composite-checkout/test/composite-checkout-thank-you.js]: Good point 👍.

File: client/lib/titan/get-eligible-titan-domain.js (hunk 2 above)
Comment (stephanethomas) [reply client/lib/titan/get-eligible-titan-domain.js]: I'll add that check 👍.

File: client/lib/titan/is-domain-eligible-for-titan-free-trial.js (hunk 1 above)
Comment (stephanethomas) [reply client/lib/titan/is-domain-eligible-for-titan-free-trial.js]: Done in https://github.com/Automattic/wp-calypso/pull/57001.

**Comment by stephanethomas** [comment]
A quick note that I'm now working on creating a `useSiteDomains()` hook. My hope is that it will allow me to keep changes to the minimum (i.e. to avoid updating all the functions to the top of the call stack in order to fetch and pass a list of domains).

File: client/lib/titan/get-eligible-titan-domain.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import { hasPaidEmailWithUs } from 'calypso/lib/emails';
+
+/**
+ * Retrieves the first domain that is eligible to Titan in this order:
+ *
+ *   - The domain from the site currently selected, if eligible
+ *   - The primary domain of the site, if eligible
+ *   - The first non-primary domain eligible found
+ *
+ * Note this method doesn't check if a domain is eligible to the 3-month free trial.
```
Comment (daledupreez) [inline client/lib/titan/get-eligible-titan-domain.js:11]: Nit: This should technically be "eligible ~to~ _for_ the 3-month free trial". (English is weird. 🙃)

I am wondering if this should also allow for the 3 month trial to be a filter. I think we're going to either need to handle this case clearly via tracking or similar, or we're going to need to not show it.

File: client/lib/titan/test/index.ts [hunk 8]
```diff
@@ -0,0 +1,144 @@
+import { getEligibleTitanDomain } from '..';
+
+describe( 'getEligibleTitanDomain()', () => {
```
Comment (daledupreez) [inline client/lib/titan/test/index.ts:3]: 😍

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts [hunk 9]
```diff
@@ -475,9 +503,61 @@ function getQuickstartUrl( {
...
+		hasTitanMail( cart )
+	) {
+		return;
+	}
+
+	if ( hasPremiumPlan( cart ) ) {
+		return;
+	}
+
+	if ( ! hasBloggerPlan( cart ) && ! hasPersonalPlan( cart ) && ! hasBusinessPlan( cart ) ) {
```
Comment (daledupreez) [inline client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:548]: I think we should show this upsell to eCommerce plan purchasers as well.
```suggestion
	if ( ! hasBloggerPlan( cart ) && ! hasPersonalPlan( cart ) && ! hasBusinessPlan( cart ) && ! hasEcommercePlan( cart ) ) {
```

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts [hunk 10]
```diff
@@ -475,9 +503,61 @@ function getQuickstartUrl( {
...
+		return;
+	}
+
+	const domain = getEligibleTitanDomain( siteSlug, domains );
```
Comment (daledupreez) [inline client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:552]: For this use case, I think we specifically want to offer the free trial, so it might be worth passing in an optional additionalFilter functional that is used to filter the list. Something like the following:
```suggestion
	const domain = getEligibleTitanDomain( siteSlug, domains, isDomainEligibleForTitanFreeTrial );
```
My thinking here is that we then _know_ that we're going to offer a free trial to the user.

I am also wondering if we need to be checking `cart` to see if it has a new domain registration using `hasNewDomainRegistration( cart )`, as that is probably the domain we should target if we can. I am honestly wondering if it makes more sense to just specify the domain name in the URL so we're only working all of this out once.

**Review by daledupreez — COMMENTED** [review]
I will test in the morning, but had the following thoughts while inspecting the code.

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts (hunk 9 above)
Comment (daledupreez) [reply client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:548]: Hmmm... We're not showing this to ECommerce plan purchasers. We likely need to update the PR description to match, as well as adding an explicit unit test to include/exclude this case.

**Review by sirbrillig — COMMENTED** [review]
I haven't tested this but I wanted to drop in and say that the code looks good to me! Thank you for updating the tests!

**Review by daledupreez — APPROVED** [review]
I've now worked through testing this and it's working as described, with the exception of showing the plan to ECommerce plans, but we can address that in a follow-up PR.

I do think we may want to add the domain name to the URL to avoid some of the current data hoops, as well as restricting the upsell to domains that are eligible for the free Professional Email trial, but those don't block us getting this deployed and we can tackle them in follow-up PRs.

### Evidence of testing

Screenshot 2021-10-15 at 17 01 59

File: client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts (hunk 9 above)
Comment (stephanethomas) [reply client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:548]: Right, I actually noticed that on Thursday but forgot to push that change 🤦‍♂️.

Comment (daledupreez) [reply client/my-sites/checkout/composite-checkout/hooks/use-get-thank-you-url/get-thank-you-page-url.ts:548]: I've updated the PR description to remove this -- we can add that in a quick follow-up PR.
